docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,340 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+    handler.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.on_error)(struct ArrowAsyncDeviceStreamHandler, int, const char*, const char*)
+
+    *Mandatory.* Handler to be called when an error is encountered by the producer. After calling
+    this, the ``release`` callback will be called as the last call on this struct. The parameters
+    are an ``errno``-compatible error code and an optional error message and metadata.
+
+    If the message and metadata are not ``NULL``, their lifetime is only valid during the scope
+    of this call. A consumer who wants to maintain these values past the return of this function
+    *MUST* copy the values themselves.
+
+    If the metadata parameter is not ``NULL``, to provide key-value error metadata, then it should
+    be encoded identically to the way that metadata is encoded in :c:member:`ArrowSchema.metadata`.
+
+    It is valid for this to be called by a producer with or without a preceding call to
+    :c:member:`ArrowAsyncProducer.request`. This callback *MUST NOT* call any methods of an
+    ``ArrowAsyncProducer`` object.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.release)(struct ArrowAsyncDeviceStreamHandler*)
+
+    *Mandatory.* A pointer to a consumer-provided release callback for the handler.
+
+    It is valid for this to be called by a producer with or without a preceding call to
+    :c:member:`ArrowAsyncProducer.request`. This must not call any methods of an ``ArrowAsyncProducer``
+    object.
+
+.. c:member:: void* ArrowAsyncDeviceStreamHandler.private_data
+
+    *Optional.* An opaque pointer to consumer-provided private data.
+
+    Producers *MUST NOT* process this member. Lifetime of this member is handled by
+    the consumer, and especially by the release callback.
+
+The ArrowAsyncTask structure
+''''''''''''''''''''''''''''
+
+The purpose of using a Task object rather than passing the array directly to the ``on_next``
+callback is to allow for more complex and efficient thread handling. Utilizing a Task
+object allows for a producer to separate the "decoding" logic from the I/O, enabling a
+consumer to avoid transferring data between CPU cores (e.g. from one L1/L2 cache to another).
+
+This producer-provided structure has the following fields:
+
+.. c:member:: int (*ArrowArrayTask.get_data)(struct ArrowArrayTask*, struct ArrowDeviceArray*)
+
+  *Mandatory.* A callback to populate the provided ``ArrowDeviceArray`` with the available data.
+  The order of ArrowAsyncTasks provided by the producer enables a consumer to know the order of
+  the data to process. If the consumer does not care about the data that is owned by this task,
+  it must still call ``get_data`` so that the producer can perform any required cleanup. ``NULL``
+  should be passed as the device array pointer to indicate that the consumer doesn't want the
+  actual data, letting the task perform necessary cleanup.
```

Comment:
Oh, it's not a pure function. Yet another reason to call this `consume_data` instead of `get_data`.